### ISSUE TITLE
bug: kafka source `scan.startup.timestamp.millis` not working as expected

### ISSUE THREAD

**Issue by stdrc** [risingwavelabs/risingwave#16046]
It's expected that RisingWave start reading from the first entry in kafka when `scan.startup.timestamp.millis` is set to a very early time, earlier than every entry in kafka. But the actually behavior is to read from offset `1`.

The root cause is that we have some chaos in the definition of `offset` in our source connector.

To reproduce, first inject the following entries into kafka:

```
{"userId": 1, "amount": 1.0, "ts": "2024-03-30 02:55:00"}
{"userId": 1, "amount": 1.0, "ts": "2024-03-30 02:55:30"}
{"userId": 1, "amount": 1.0, "ts": "2024-03-30 02:55:40"}
```

Then run:

```sql
create source test_src (
    userId int,
    amount decimal,
    ts timestamp,
    watermark for ts as ts - interval '15 second'
)
with (
    connector = 'kafka'
    , topic = 'test'
    , properties.bootstrap.server = '127.0.0.1:29092'
    , scan.startup.timestamp_millis = '140000000000'
) format plain encode json;

create materialized view mv_simple as
select * from test_src;

select * from test_src;
select * from mv;
```

The last two queries are expected to return same dataset but not, the first entry is missing in the result of the last query.

**Comment by BugenZhao** [comment]
Can refer to https://risingwave-labs.slack.com/archives/C03A2PSS8KU/p1711437568241809

**Comment by stdrc** [comment]
It seems the concept of `offset` is a little bit chaos in our source. I will do some refactor later, for now let's do a quick fix first (#16073).
